tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/aggregates.md

```diff
@@ -0,0 +1,183 @@
+    break;
+// ...
+```
+
+Now let's update our generated SQL to use the generated target list:
+
+```typescript
+// ...
+const sql = `SELECT ${target_list.join(", ")} FROM (
+               (
+                 SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
+               )`;
+// ...
+```
+
+Note that we form the set of rows to be aggregated first, so that the limit and offset clauses are applied correctly.
+
+And instead of returning all rows, we're going to assume that we only get a single row back, so we can match on that and
+return the single row of aggregates:
+
+```typescript
+const result = await state.db.get(sql, ...parameters);
+
+if (result === undefined) {
+    throw new InternalServerError("Unable to fetch aggregates");
+}
+
+return result;
+```
+
+Here's the full function:
+
+```typescript
+async function fetch_aggregates(state: State, request: QueryRequest): Promise<{
+    [k: string]: unknown
+}> {
+    const target_list = [];
+
+    for (const aggregateName in request.query.aggregates) {
+        if (Object.prototype.hasOwnProperty.call(request.query.aggregates, aggregateName)) {
+            const aggregate = request.query.aggregates[aggregateName];
+            switch(aggregate.type) {
+                case 'star_count':
+                    target_list.push(`COUNT(1) AS ${aggregateName}`);
+                    break;
+                case 'column_count':
+                    target_list.push(`COUNT(${aggregate.distinct ? 'DISTINCT ' : ''}${aggregate.column}) AS ${aggregateName}`);
+                    break;
+                case 'single_column':
+                    throw new NotSupported("custom aggregates not yet supported");
+            }
+        }
+    }
+
+    const parameters: any[] = [];
+
+    const limit_clause = request.query.limit == null ? "" : `LIMIT ${request.query.limit}`;
+    
+    const offset_clause = request.query.offset == null ? "" : `OFFSET ${request.query.offset}`;
+
+    const where_clause = request.query.where == null ? "" : `WHERE ${visit_expression(parameters, request.query.where)}`;
+
+    const order_by_clause = request.query.order_by == null ? "" : `ORDER BY ${visit_order_by_elements(request.query.order_by.elements)}`;
+
+    const sql = `SELECT ${target_list.join(", ")} FROM (
+                    SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
+                )`;
+
+    console.log(JSON.stringify({ sql, parameters }, null, 2));
+
+    const result = state.db.get(sql, ...parameters);
+
+    if (result === undefined) {
+        throw new InternalServerError("Unable to fetch aggregates");
+    }
+
+    return result;
+}
+```
+
+That's it, so let's test our connector one more time, and hopefully see some passing tests this time. And let's 
+check that we're generating the right SQL.
```

Comment:
I didn't mention this on the `predicates` page, but let's be explicit and tell them they need to kill and recompile their connector before running the tests.